File: app/src/main/java/xyz/flove/square/Board.java
package xyz.flove.square;

import java.util.ArrayList;

import xyz.flove.square.entities.Color;
import xyz.flove.square.entities.Position;

/**
 * Created by liulnn on 17/4/19.
 */

public class Board implements Cloneable {
    public Status status;
    private Piece[][] panel;
    private int removeCount;

    public Board(int size) {
        panel = new Piece[size][size];
        status = Status.DOWN;
        removeCount = 0;
    }

    enum Status {
        // 下子阶段，提子阶段，对战阶段（吃子）
        DOWN, REMOVE, FIGHT, EAT
    }

    public boolean isSquare(Position position) {
        Piece piece = getPiece(position);
        if (piece != null && piece.squares != null && piece.squares.length > 0) {
            return true;
        }
        return false;
    }

    public int getLength() {
        return panel[0].length;
    }

    public Piece getPiece(Position position) {
        return panel[position.x][position.y];
    }

    public Piece getPiece(int x, int y) {
        return panel[x][y];
    }

    public void addPiece(Position position, String color) {
        addPiece(new Piece(position, color));
    }

    public boolean addPiece(Piece piece) {
        if (panel[piece.position.x][piece.position.y] != null) {
            return false;
        }
        panel[piece.position.x][piece.position.y] = piece;
        updateStatus();
        return true;
    }


    public void removePiece(Position position, String color) {
        removePiece(new Piece(position, color));
    }

    public boolean removePiece(Piece piece) {
        if (panel[piece.position.x][piece.position.y] == null
                || !panel[piece.position.x][piece.position.y].color.equals(piece.color)) {
            return false;
        }
        panel[piece.position.x][piece.position.y] = null;
        removeCount += 1;
        if (removeCount >= 2) {
            status = Board.Status.FIGHT;
        }
        return true;
    }

    private void updateStatus() {
        if (status == Status.DOWN) {
            boolean isFull = true;
            for (int i = 0; i < getLength(); i++) {
                for (int j = 0; j < getLength(); j++) {
                    if (panel[i][j] == null || panel[i][j].color.equals(Color.NULL)) {
                        isFull = false;
                    }
                }
            }
            if (isFull) {
                status = Status.REMOVE;
            }
        }
    }

    public Position[] getCanEatPieces(String color){
        ArrayList<Position> positions = new ArrayList<>();
        for (int i = 0; i < getLength(); i++) {
            for (int j = 0; j < getLength(); j++) {
                if(panel[i][j] != null && !panel[i][j].color.equals(color) && !isSquare(new Position(i, j))){
                    positions.add(panel[i][j].position);
                }
            }
        }
        return positions.toArray(new Position[positions.size()]);
    }

    public void draw() {
        System.out.println("--------------------------------");
        for (int i = 0; i < getLength(); i++) {
            for (int j = 0; j < getLength(); j++) {
                String color = Color.NULL;
                if (panel[j][i] != null) {
                    color = panel[j][i].color;
                }
                if (color.equals(Color.BLACK)) {
                    System.out.print("[*]");
                } else if (color.equals(Color.WHITE)) {
                    System.out.print("[#]");
                } else {
                    System.out.print("[ ]");
                }
            }
            System.out.println("");
        }
        System.out.println("--------------------------------");
        for (int i = 0; i < getLength(); i++) {
            for (int j = 0; j < getLength(); j++) {
                if (this.isSquare(new Position(j, i))) {
                    System.out.print("[+]");
                } else {
                    System.out.print("[ ]");
                }
            }
            System.out.println("");
        }
    }

    @Override
    public Board clone() {
        Board newBoard = null;
        try {
            newBoard = (Board) super.clone();
            Piece[][] pieces = new Piece[getLength()][getLength()];
            for (int i = 0; i < getLength(); i++) {
                for (int j = 0; j < getLength(); j++) {
                    if (panel[i][j] != null) {
                        pieces[i][j] = new Piece(new Position(i, j), panel[i][j].color);
                    }
                }
            }
            newBoard.panel = pieces;
            newBoard.status = Status.valueOf(status.name());
            return newBoard;
        } catch (CloneNotSupportedException e) {
        } finally {
            return newBoard;
        }
    }
}

File: app/src/main/java/xyz/flove/square/Board.kt
package xyz.flove.square

import xyz.flove.square.core.*


class Board(var panel: Array<Array<Piece?>>) : Cloneable {
    var size: Int = panel.size
    var status: BoardStatus? = null
    private var removeCount: Int = 0

    init {
        status = BoardStatus.DOWN
        removeCount = 0
    }


    fun isSquare(piece: Piece): Boolean {
        if (piece.squares.isNotEmpty()) {
            return true
        }
        return false
    }

    fun isSquare(position: Position): Boolean {
        val piece = getPiece(position)
        if (piece != null) {
            return isSquare(piece)
        }
        return false
    }

    fun getMoveDirection(position: Position): List<Direction> {
        val directions: MutableList<Direction> = ArrayList()
        val piece = getPiece(position) ?: return directions
        if (piece.position.x - 1 >= 0) {
            val westPiece = getPiece(position.x - 1, position.y)
            if (westPiece == null) {
                directions.add(Direction.WEST)
            }
        }
        if (piece.position.y - 1 >= 0) {
            val northPiece = getPiece(position.x, position.y - 1)
            if (northPiece == null) {
                directions.add(Direction.NORTH)
            }
        }
        if (piece.position.x + 1 <= 4) {
            val eastPiece = getPiece(position.x + 1, position.y)
            if (eastPiece == null) {
                directions.add(Direction.EAST)
            }
        }
        if (piece.position.y + 1 <= 4) {
            val sorthPiece = getPiece(position.x, position.y + 1)
            if (sorthPiece == null) {
                directions.add(Direction.SORTH)
            }
        }
        return directions
    }


    val length: Int
        get() = this.size

    fun getPiece(position: Position): Piece? {
        return panel[position.x][position.y]
    }

    fun getPiece(x: Int, y: Int): Piece? {
        return panel[x][y]
    }

    fun addPiece(position: Position, color: Color): Boolean {
        var piece: Piece = Piece(position, color)
        return addPiece(piece)
    }

    fun addPiece(piece: Piece): Boolean {
        if (panel[piece.position.x][piece.position.y] != null) {
            return false
        }
        panel[piece.position.x][piece.position.y] = piece
        updateStatus()
        return true
    }

    fun removePiece(piece: Piece): Boolean {
        if (panel[piece.position.x][piece.position.y] == null
                || panel[piece.position.x][piece.position.y]!!.color != piece.color) {
            return false
        }
        panel[piece.position.x][piece.position.y] = null
        removeCount += 1
        if (removeCount >= 2 && status == BoardStatus.REMOVE) {
            status = BoardStatus.FIGHT
        }
        return true
    }

    private fun updateStatus() {
        if (status == BoardStatus.DOWN) {
            if (isFull()) {
                status = BoardStatus.REMOVE
            }
        }
    }


    fun isFull(): Boolean {
        for (i in 0..length - 1) {
            panel[i].asList().forEach {
                if (it == null || it.color == Color.NULL) return false
            }
        }
        return true
    }


    fun getCanEatPieces(color: Color): List<Position> {
        val positions: MutableList<Position> = ArrayList()
        for (i in 0..length - 1) {
            panel[i].forEach {
                if (it != null && it.color != color && !isSquare(it))
                    positions.add(it.position)
            }
        }
        return positions
    }

    fun isLazy(color: Color): Boolean {
        return false
    }

    fun getResult(color: Color): Result {
        if (status == BoardStatus.DOWN || status == BoardStatus.REMOVE) {
            return Result.NULL
        }
        var enemyPieceCount = 0
        var enemyCanMovePieceCount = 0
        for (i in 0..length - 1) {
            for (j in 0..length - 1) {
                val piece = panel[i][j]
                val directions = getMoveDirection(Position(i, j))
                if (piece != null && piece.color != color) {
                    enemyPieceCount += 1
                    if (directions.isNotEmpty()) {
                        enemyCanMovePieceCount += 1
                    }
                }
            }
        }
        //对方棋子少于3个
        if (enemyPieceCount < 3) {
            return Result.WINNER
        }
        //对方没有可以移动的棋子
        if (enemyCanMovePieceCount == 0) {
            return Result.WINNER
        }

        //己方是否是懒棋
        if (isLazy(color)) {
            return Result.LOSER
        }
        return Result.NULL
    }

    fun draw() {
        println("--------------------------------")
        for (i in 0..length - 1) {
            for (j in 0..length - 1) {
                var color = Color.NULL
                if (panel[j][i] != null) {
                    color = panel[j][i]!!.color
                }
                if (color == Color.BLACK) {
                    print("[*]")
                } else if (color == Color.WHITE) {
                    print("[#]")
                } else {
                    print("[ ]")
                }
            }
            println("")
        }
        println("--------------------------------")
        for (i in 0..length - 1) {
            for (j in 0..length - 1) {
                if (this.isSquare(Position(j, i))) {
                    print("[+]")
                } else {
                    print("[ ]")
                }
            }
            println("")
        }
    }

    public override fun clone(): Board {
        var newBoard: Board = this
        val pieces: Array<Array<Piece?>> = Array(length) { Array<Piece?>(length, { null }) }
        try {
            newBoard = super.clone() as Board
            for (i in 0..length - 1) {
                for (j in 0..length - 1) {
                    if (panel[i][j] != null) {
                        pieces[i][j] = Piece(Position(i, j), panel[i][j]!!.color)
                    }
                }
            }
            newBoard.panel = pieces
            newBoard.status = BoardStatus.valueOf(status!!.name)
        } catch (e: CloneNotSupportedException) {
        } finally {
            return newBoard
        }
    }
}